src/mongoc/mongoc-bulk-operation.c

```diff
@@ -851,6 +851,16 @@ mongoc_bulk_operation_set_client (mongoc_bulk_operation_t *bulk, void *client)
 }
 
 
+void
+mongoc_bulk_operation_set_client_session (mongoc_bulk_operation_t *bulk,
+                                          void *client_session)
+{
+   BSON_ASSERT (bulk);
+
+   bulk->session = (mongoc_client_session_t *) client_session;
```

Comment:
I realized this makes it possible to assign a session belonging to a different client. A few options to address this:

 * Change this function to accept a `bson_error_t *` and return a success boolean
 * `BSON_ASSERT()` that `client_session->client` is the same as `bulk->client`

Either way, I think we also need to modify `mongoc_bulk_operation_set_client()` to unset the `client_session`. Otherwise, it'd be possible the make a valid session assignment and then screw it up by changing the client again.

Thoughts?

Reply:
Let's BSON_ASSERT that client_session->client is the same as bulk->client, in both mongoc_bulk_operation_set_client and mongoc_bulk_operation_set_session, so the condition is checked when you set them in either order.